### ISSUE TITLE
Read Replicas

### ISSUE THREAD

**Issue by knpwrs** [gajus/slonik#22]
Does slonik have any way or would there be interest in implementing an abstraction over sending read queries to read replicas and write queries to a master server? I know you aren't a fan of ORMs (based on the medium article) but this is a feature which happens to be supported by Sequelize: http://docs.sequelizejs.com/manual/installation/usage.html#read-replication

**Comment by gajus** [comment]
The last I checked, this is a very complicated issue.

From the top of my head, some considerations would include:

* All transactions would automatically need to use write replica.
* All queries that use functions would almost certainly need to use write replica. Perhaps a whitelist would help, but thats getting into a dangerous territory.

**Comment by knpwrs** [comment]
Maybe allowing middleware to manage pools of connections could be useful? Users could create middleware for their own query balancing logic and slonik could ship with a basic implementation that covers certain use cases.

**Comment by gajus** [comment]
> Maybe allowing middleware to manage pools of connections could be useful? Users could create middleware for their own query balancing logic and slonik could ship with a basic implementation that covers certain use cases.

In general, this is smart. Just needs some thinking how to implement it.

Problem is that most of the time you already have a connection ready before making a query (with exception of executing queries using the connection pool object directly).

If we focus only the use case for when the connection has not been retrieved from the pool, then this is simple to achieve.

I am looking at our code base and I cannot see many examples where it would be safe though/ reasonable/ worth the complexity.

Do you have any examples that you could share?

Just structure/ problem. No need to share actual queries.

**Comment by knpwrs** [comment]
The queries in my current project are all very simple so I'm able to get away with just checking if a query is an update/insert/delete so I can route it to the appropriate pool. I'm building on top of [`node-postgres`](https://github.com/brianc/node-postgres) and [`sql-template-strings`](https://github.com/felixfbecker/node-sql-template-strings) to make the following tagged template literal which queries the database and returns the result:

```ts
export const query = async <T>(strings: TemplateStringsArray, ...values: any[]): Promise<T[]> => {
  const stmt = sql(strings, ...values); // sql-template-strings
  const conn = await (isWrite(stmt.sql) ? writePool : readPool).connect();
  try {
    const res = await conn.query(stmt);
    return getData<T>(res);
  } finally {
    conn.release();
  }
};
```

Example usage:

```ts
const [user] = await query<User>`SELECT * FROM user WHERE id = ${userId}`;
```

`getData` handles `camelCasing` and whatnot.

I also have my own solution for transactions which essentially passes a callback function a scoped tagged template literal and automatically handles calling `COMMIT` and `ROLLBACK`.

Basically I'm not actually using slonik as of yet, but I stumbled across it this morning and I noticed some overlap in abstractions that I've been writing so I thought it might be nice to use an open-source solution rather than maintaining my own abstractions, but I want to be able to maintain my existing query-balancing functionality. If I'm understanding what you're asking then essentially the answer is that my tagged template literal handles getting a connection from the pool -- I don't have anything set up where I need to get and then use a connection.

**Comment by gajus** [comment]
This sounds to me like it requires another layer of abstraction, e.g.

```js
import {
  createPool,
  createLazyPool
} from 'slonik';

const pool = createLazyPool([
  {
    name: 'READ',
    pool: createPool(),
    matchQuery: () => { /* logic to match the query */ }
  },
  
  // Always fallback to the last pool.
  {
    name: 'WRITE',
    pool: createPool()
  }
]);

// Proceed to use Slonik as you would.
pool;

```

Not going to rush this one.

Want to think of all possible ways this can backfire.

**Comment by gajus** [comment]
@knpwrs https://github.com/gajus/slonik#routing-queries-to-different-connections

This should solve the problem!

**Comment by knpwrs** [comment]
Nice! Thanks!
